### PR TITLE
Make redirect middleware can use path

### PR DESCRIPTION
**Pull request by takmatsu**
This patch enable us to use path in redirect middleware.

Currently, redirect middleware respects only scheme and host.
But actually our S3-Compatible storage needs this features for redirect to object storage endpoint.
and I believe this should be useful also for other users.

Related: #3022

### CODE REVIEW

File: registry/storage/driver/middleware/redirect/middleware.go [hunk 1]
```diff
@@ -37,10 +38,13 @@ func newRedirectStorageMiddleware(sd storagedriver.StorageDriver, options map[st
 		return nil, fmt.Errorf("no host specified for redirect baseurl")
 	}
 
-	return &redirectStorageMiddleware{StorageDriver: sd, scheme: u.Scheme, host: u.Host}, nil
+	return &redirectStorageMiddleware{StorageDriver: sd, scheme: u.Scheme, host: u.Host, path: u.Path}, nil
 }
 
 func (r *redirectStorageMiddleware) URLFor(ctx context.Context, path string, options map[string]interface{}) (string, error) {
+	if r.path != "" {
+		path = r.path + path
```
Comment (waynr) [inline registry/storage/driver/middleware/redirect/middleware.go:46]: We should use `path.Join` here to ensure that the `r.path` and `path` parts are separated by slashes even if the user neglects to include those slashes. We should also add some test cases to validate the new behavior of `URLFor` when `path` is set (including `path` and `r.path` values with and without leading/trailing slashes to validate proper path joining behavior.

File: registry/storage/driver/middleware/redirect/middleware.go [hunk 2]
```diff
@@ -13,6 +13,7 @@ type redirectStorageMiddleware struct {
 	storagedriver.StorageDriver
 	scheme string
 	host   string
+	path   string
```
Comment (waynr) [inline registry/storage/driver/middleware/redirect/middleware.go:16]: This should be named `basePath`.

**Comment by alibowl** [comment]
Any reason this hasn't been merged yet, we'd like to use this feature

**Comment by rikatz** [comment]
@milosgajdos 👋 
Thanks for the project!

I was looking into this very same feature and got hit by the same problem. Apparently, the path is not respected :( Is there anything that can be done here to get this PR merged, or should people willing on having this same middleware feature use this as an extension point on docker distribution but write their own middleware?

Thank you very much!!!

**Comment by milosgajdos** [comment]
I'll close and reopen this PR so it re-kicks the CI.

File: registry/storage/driver/middleware/redirect/middleware.go [hunk 3]
```diff
@@ -4,16 +4,17 @@ import (
 	"context"
 	"fmt"
 	"net/url"
+	pathutil "path"
```
Comment (milosgajdos) [inline registry/storage/driver/middleware/redirect/middleware.go:7]: is there a valid reason why we're aliasing `path` import to `pathutil`?

**Comment by milosgajdos** [comment]
@takmatsu can you also please squash the commits

**Review by milosgajdos — APPROVED** [review]

**Comment by rikatz** [comment]
@milosgajdos thank you very very much for pushing this :))

**Pull request merged**